# DSDBAL ASSIGNMENT NO. 3

## 1. Importing all the required Python Libraries.

In [394]:
import numpy as np
import pandas as pd

## 2. Loading the Dataset into pandas’ data frame.

In [395]:
df=pd.read_csv('nba.csv')

In [396]:
df.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

Datatypes like Number ,Age and Weight are in Float that can be converted to Int datatype as all have 
Height has (a-b) style that can be rewritten as 6.2 

In [397]:
df.shape

(458, 9)

The dataset has a total of 458 rows and 9 attributes 

## 3.Handling Missing Values

In [398]:
df.isnull().sum()

Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary      12
dtype: int64

In [399]:
for col in df.columns:
    if df[col].dtype=="float64":  # Check if column is float type
        col_mean = df[col].mean()  # Calculate mean
        df[col] = df[col].fillna(col_mean)  
df.isnull().sum()

Name         1
Team         1
Number       0
Position     1
Age          0
Height       1
Weight       0
College     85
Salary       0
dtype: int64

In [400]:
for col in df.columns:
    if df[col].dtype=='object':
        df[col]=df[col].fillna(df[col].mode()[0])

In [401]:
df.isnull().sum()

Name        0
Team        0
Number      0
Position    0
Age         0
Height      0
Weight      0
College     0
Salary      0
dtype: int64

## 4.Applying Proper Type Conversions

In [402]:
columns_to_convert = ["Number", "Age", "Weight","Salary"]  # List of columns to convert

for col in columns_to_convert:
    df[col]=df[col].astype(int)

In [403]:
df.dtypes

Name        object
Team        object
Number       int32
Position    object
Age          int32
Height      object
Weight       int32
College     object
Salary       int32
dtype: object

In [404]:
for col in df.columns:
    if df[col].dtype=='object':
        df[col]=df[col].astype('string')
df.dtypes

Name        string[python]
Team        string[python]
Number               int32
Position    string[python]
Age                  int32
Height      string[python]
Weight               int32
College     string[python]
Salary               int32
dtype: object

### Height attribute is in string type as 6-9,6-11,therefore converting it to float 

In [405]:
df['Height'] = df['Height'].fillna(df['Height'].mode()[0])

In [406]:
def convert_height(height):
    height = str(height)  # Convert to string if it's not already
    return float(height.replace('-', '.'))

df['Height']=df['Height'].apply(convert_height)
df.dtypes

Name        string[python]
Team        string[python]
Number               int32
Position    string[python]
Age                  int32
Height             float64
Weight               int32
College     string[python]
Salary               int32
dtype: object

In [407]:
df.dtypes

Name        string[python]
Team        string[python]
Number               int32
Position    string[python]
Age                  int32
Height             float64
Weight               int32
College     string[python]
Salary               int32
dtype: object

In [408]:
df.head()

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25     6.2     180   
1    Jae Crowder  Boston Celtics      99       SF   25     6.6     235   
2   John Holland  Boston Celtics      30       SG   27     6.5     205   
3    R.J. Hunter  Boston Celtics      28       SG   22     6.5     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29     6.1     231   

             College   Salary  
0              Texas  7730337  
1          Marquette  6796117  
2  Boston University  4842684  
3      Georgia State  1148640  
4           Kentucky  5000000

## 5. Central Tendencies - Measures

### 5.1.1. Mean (Built-in method)

In [409]:
for col in df.columns:
    if df[col].dtype=='int'or df[col].dtype=='float':
        print(f" Mean value of {col}: {df[col].mean()}")
        # print(f" Median value of {col}: {df[col].median()}")
        print("-"*40)


 Mean value of Number: 17.676855895196507
----------------------------------------
 Mean value of Age: 26.936681222707424
----------------------------------------
 Mean value of Height: 6.517969432314411
----------------------------------------
 Mean value of Weight: 221.52183406113537
----------------------------------------
 Mean value of Salary: 4842684.102620088
----------------------------------------


### 5.1.2. Mean (User-defined method)

In [410]:
mean={}
for col in df.columns:
    if df[col].dtype in ['int','float64']:
        total=0
        count=0
        for value in df[col]:
            if pd.notna(value):
                total=total+value
                count=count+1
        means[col]=total/count if count!=0 else None

for i,j in means.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  17.676855895196507
----------------------------------------
Age  :  26.936681222707424
----------------------------------------
Height  :  6.517969432314407
----------------------------------------
Weight  :  221.52183406113537
----------------------------------------
Salary  :  4842684.102620088
----------------------------------------


### 5.2.1. Median (Built-in method)

In [411]:
for col in df.columns:
    if df[col].dtype=='int'or df[col].dtype=='float':
        print(f" Median value of {col}: {df[col].median()}")
        print("-"*40)

 Median value of Number: 13.0
----------------------------------------
 Median value of Age: 26.0
----------------------------------------
 Median value of Height: 6.6
----------------------------------------
 Median value of Weight: 220.0
----------------------------------------
 Median value of Salary: 2880600.0
----------------------------------------


### 5.2.2. Median (User-defined method)

In [412]:
medians={}

for col in df.columns:
    if df[col].dtype in ['int','float64']:
        values=sorted([x for x in df[col]])
        n=len(values)
        if n==0:
            medians[col]=None
        elif n%2==1:
            medians[col]=values[n//2]
        else:
            medians[col]=(values[n//2-1]+values[n//2])/2

for i,j in medians.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  13.0
----------------------------------------
Age  :  26.0
----------------------------------------
Height  :  6.6
----------------------------------------
Weight  :  220.0
----------------------------------------
Salary  :  2880600.0
----------------------------------------


### 5.3.1. Mode (Built-in method)

In [413]:
for col in df.columns:
    # if df[col].dtype=='string':
    print(f" Mode value of {col}: {df[col].mode()[0]}")
    print("-"*153)

 Mode value of Name: Aaron Brooks
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Mode value of Team: New Orleans Pelicans
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Mode value of Number: 5
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Mode value of Position: SG
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Mode value of Age: 24
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Mode value of Height: 6.1
---------------------------------------------------

### 5.3.2. Mode (User-defined method)

In [414]:
modes = {}
from collections import Counter

for col in df.columns:
    values = [x for x in df[col] if pd.notna(x)]  
    if values:  
        most_common = Counter(values).most_common()  
        max_freq = most_common[0][1]  
        mode_values = [key for key, val in most_common if val == max_freq]
        modes[col] = mode_values[0] if len(mode_values) == 1 else mode_values
    else:
        modes[col] = None 

for i,j in modes.items():
    print(i," : " ,j)
    print("-"*40)

Name  :  Aaron Brooks
----------------------------------------
Team  :  New Orleans Pelicans
----------------------------------------
Number  :  5
----------------------------------------
Position  :  SG
----------------------------------------
Age  :  24
----------------------------------------
Height  :  6.1
----------------------------------------
Weight  :  220
----------------------------------------
College  :  Kentucky
----------------------------------------
Salary  :  947276
----------------------------------------


### 5.4.1 Maximum value (Built-in method)

In [415]:
# print(df['Height'].max())
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f'Maximum value in {col} is {df[col].max()}')
        print("-"*40)

Maximum value in Number is 99
----------------------------------------
Maximum value in Age is 40
----------------------------------------
Maximum value in Height is 7.3
----------------------------------------
Maximum value in Weight is 307
----------------------------------------
Maximum value in Salary is 25000000
----------------------------------------


### 5.4.2 Maximum value (User-defined method)

In [416]:
maximum={}
maxval=0

for col in df.columns:
    if df[col].dtype in ['int','float']:
        values=[x for x in df[col]]
        if values:
            maxval=values[0]
            for i in values:
                if i>maxval:
                    maxval=i
            maximum[col]=maxval
        else:
            maximum[col]=None  

for i,j in maximum.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  99
----------------------------------------
Age  :  40
----------------------------------------
Height  :  7.3
----------------------------------------
Weight  :  307
----------------------------------------
Salary  :  25000000
----------------------------------------


### 5.5.1 Minimum value (Built-in method)

In [417]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f'Minimum value in {col} is {df[col].min()}')
        print("-"*40)

Minimum value in Number is 0
----------------------------------------
Minimum value in Age is 19
----------------------------------------
Minimum value in Height is 5.11
----------------------------------------
Minimum value in Weight is 161
----------------------------------------
Minimum value in Salary is 30888
----------------------------------------


### 5.5.2 Minimum value ( User-defined method)

In [418]:
minimum={}
minval=0

for col in df.columns:
    if df[col].dtype in ['int','float']:
        values=[x for x in df[col]]
        if values:
            minval=values[0]
            for i in values:
                if i<minval:
                    minval=i
            minimum[col]=minval
        else:
            minimum[col]=None  

for i,j in minimum.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  0
----------------------------------------
Age  :  19
----------------------------------------
Height  :  5.11
----------------------------------------
Weight  :  161
----------------------------------------
Salary  :  30888
----------------------------------------


### 5.6.1. Standard Deviation (Built-in method)

In [419]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f"Standard Deviation of {col}: {df[col].std()}")

Standard Deviation of Number: 15.94864396617882
Standard Deviation of Age: 4.399414063696201
Standard Deviation of Height: 0.35227429750526434
Standard Deviation of Weight: 26.339488773302435
Standard Deviation of Salary: 5160125.694003381


### 5.6.2. Standard Deviation (User-defined method)

In [420]:
std_dev_values = {}

for col in df.columns:
    if df[col].dtype in ['int', 'float64']:
        values = [x for x in df[col] if pd.notna(x)] 
        n = len(values)
        if n > 1:  
            mean_value = sum(values) / n 
            squared_diffs = [(x - mean_value) ** 2 for x in values]
            variance = sum(squared_diffs) / (n - 1)
            std_dev = variance ** 0.5  
            std_dev_values[col] = std_dev  
        else:
            std_dev_values[col] = None 
for i,j in std_dev_values.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  15.948643966178828
----------------------------------------
Age  :  4.399414063696205
----------------------------------------
Height  :  0.3522742975052651
----------------------------------------
Weight  :  26.33948877330243
----------------------------------------
Salary  :  5160125.694003384
----------------------------------------


### 5.7.1. Range (Built-in method)

In [421]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f'Minimum value in {col} is {(df[col].max())-(df[col].min())}')
        print("-"*40)

Minimum value in Number is 99
----------------------------------------
Minimum value in Age is 21
----------------------------------------
Minimum value in Height is 2.1899999999999995
----------------------------------------
Minimum value in Weight is 146
----------------------------------------
Minimum value in Salary is 24969112
----------------------------------------


In [422]:
range_values = {}

for col in df.columns:
    # Check if the column is numeric (int or float)
    if df[col].dtype in ['int64', 'float64']:
        values = df[col].dropna()  # Remove NaN values
        range_values[col] = max(values) - min(values)  # Use built-in functions

print(range_values)

{'Height': 2.1899999999999995}
